batch_job/src/batch_job_amazon_batch.c

```diff
@@ -0,0 +1,639 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+#include "batch_job_internal.h"
+#include "process.h"
+#include "batch_job.h"
+#include "stringtools.h"
+#include "debug.h"
+#include "jx_parse.h"
+#include "jx.h"
+#include "jx_pretty_print.h"
+#include "itable.h"
+#include "hash_table.h"
+
+
+#include <time.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+static int initialized = 0;
+static char* queue_name = NULL;
+static char* compute_env_name = NULL;
+static char* vpc = NULL;
+static char* sec_group = NULL;
+static char* subnet = NULL;
+//static int ids = 1;
+static struct itable* done_jobs;//might need an itable of itables for different runs....
+static struct itable* amazon_job_ids;
+static struct itable* done_files;
+static int instID;
+static char* bucket_name = NULL;
+static struct hash_table* submitted_files = NULL;
+static int HAS_SUBMITTED_VALUE = 1;
+
+union amazon_batch_ccl_guid{
+	char c[8];
+	unsigned int ul;
+};
+
+static struct internal_amazon_batch_amazon_ids{
+	char* aws_access_key_id;
+	char* aws_secret_access_key;
+	char* aws_region;
+	char* aws_email;
+	char* master_env_prefix;
+}initialized_data;
+
+static unsigned int gen_guid(){
+	FILE* ran = fopen("/dev/urandom","r");
+	if(!ran)
+		fatal("Cannot open /dev/urandom");
+	union amazon_batch_ccl_guid guid;
+	size_t k = fread(guid.c,sizeof(char),8,ran);
+	if(k<8)
+		fatal("couldn't read 8 bytes from /dev/urandom/");
+	fclose(ran);
+	return guid.ul;	
+}
+static struct jx* run_command(char* cmd){
+	FILE* out = popen(cmd,"r");
+	struct jx* jx = jx_parse_stream(out);
+	pclose(out);
+	return jx;
+}
+
+static void split_comma_list(char* in, int* size, char*** output){
+	*size = 0;
+	char* tmp = strdup(in);
+	char* ta = strtok(tmp,",");
+	while(ta != NULL){
+		if(strlen(ta) < 1) continue;
+		*size += 1;
+		//free(ta);//hoepfully doesn't crash everything: it did...
+		ta = strtok(0,",");
+	}
+	
+	*output = malloc(sizeof(char*) * (*size));
+	int i = 0;
+	tmp = strdup(in);
+	ta = strtok(tmp,",");
+	while(ta != NULL){
+		if(strlen(ta) < 1) continue;
+		(*output)[i] = strdup(ta);
+		ta = strtok(0,",");
+		i+=1;
+	}
+	
+	return;
+	
+}
+
+static void clean_str_array(int size, char*** array){
+	int i=0;
+	for(i=0; i< size; i++){
+		free((*array)[i]);
+	}
+	free(*array);
+}
+
+static void upload_input_files_to_s3(char* files,char* jobname){
```

Comment:
Uploads can and will fail. This function must detect those failures and return overall success or failure.